mllib/src/main/scala/org/apache/spark/ml/feature/StratifiedSampling.scala

```diff
@@ -0,0 +1,129 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.ml.feature
+
+import org.apache.hadoop.fs.Path
+
+import org.apache.spark.annotation.{Experimental, Since}
+import org.apache.spark.ml.Transformer
+import org.apache.spark.ml.feature.StratifiedSampling.StratifiedSamplingWriter
+import org.apache.spark.ml.param.ParamMap
+import org.apache.spark.ml.param.shared._
+import org.apache.spark.ml.util._
+import org.apache.spark.sql.{DataFrame, Row}
+import org.apache.spark.sql.types.{StringType, StructType}
+import org.apache.spark.util.Utils
+
+/**
+ * :: Experimental ::
+ *
+ * Stratified sampling on the DataFrame according to the keys in a specific label column. User
+ * can set 'fraction' to set different sampling rate for each key.
+ *
+ * @param withReplacement can elements be sampled multiple times (replaced when sampled out)
+ * @param fraction expected size of the sample as a fraction of the items
+ *  without replacement: probability that each element is chosen; fraction must be [0, 1]
+ *  with replacement: expected number of times each element is chosen; fraction must be >= 0
+ */
+@Experimental
+final class StratifiedSampling private(
+    override val uid: String,
+    val withReplacement: Boolean,
+    val fraction: Map[String, Double])
+  extends Transformer with HasLabelCol with HasSeed with DefaultParamsWritable {
+
+  @Since("2.0.0")
+  def this(withReplacement: Boolean, fraction: Map[String, Double]) =
+    this(Identifiable.randomUID("stratifiedSampling"), withReplacement, fraction)
+
+  /** @group setParam */
+  @Since("2.0.0")
+  def setSeed(value: Long): this.type = set(seed, value)
+
+  /** @group setParam */
+  @Since("2.0.0")
+  def setLabel(value: String): this.type = set(labelCol, value)
+
+  setDefault(seed -> Utils.random.nextLong)
```

Comment:
This is not necessary. We assign a fixed seed based on the hash value of the class name.